### DISCUSSION TITLE
Create plugin: addBase does not work inside of matchComponents

### DISCUSSION THREAD

**Discussion by MatthiasDunker** [tailwindlabs/tailwindcss#13752] [Help]
**What version of Tailwind CSS are you using?**

3.4.3

**What build tool (or framework if it abstracts the build tool) are you using?**

Postcss 8.4.38 in Next.js 14.1.0

**What version of Node.js are you using?**

v20.5.0

**What browser are you using?**

Chrome

**What operating system are you using?**

macOS

**Reproduction URL**

[Codesandbox.io](https://codesandbox.io/p/devbox/funny-hoover-forked-hcg5m6?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clwomhmg800063b6odgnlkrh4%2522%252C%2522sizes%2522%253A%255B57.38379814077025%252C42.61620185922975%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clwomhmg800023b6o5beisu67%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clwomhmg800043b6o7exkrhl1%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clwomhmg800053b6o40v0agjl%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clwomhmg800023b6o5beisu67%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwomhmg800013b6oyuqf55q5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clwomhmg800023b6o5beisu67%2522%252C%2522activeTabId%2522%253A%2522clwomhmg800013b6oyuqf55q5%2522%257D%252C%2522clwomhmg800053b6o40v0agjl%2522%253A%257B%2522id%2522%253A%2522clwomhmg800053b6o40v0agjl%2522%252C%2522activeTabId%2522%253A%2522clwomty2c01373b6oq6yfjrcm%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clwomty2c01373b6oq6yfjrcm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clwomhmg800043b6o7exkrhl1%2522%253A%257B%2522id%2522%253A%2522clwomhmg800043b6o7exkrhl1%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwomhmg800033b6oml6azzfy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clwomhmur000hd9f8fshg248m%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522clwomtsn7010x3b6ouzod8kwk%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clwomtsn7010x3b6ouzod8kwk%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Describe your issue**

As you can see, classes `typo-headline` and `typo-body` get rendered, but `size-small`
 does not.
 
 The nested pattern of plugin methods does not seem to work. But I would expect `addBase` to create a `size-small` class.
 
 ```
matchComponents(
        {
          typo: (value) => {
            console.log("🔴", value);

            addBase({
              ".size-small": {
                [`--typo-${value}-fz`]: "1.5rem",
              },
            });

            return [
              {
                fontSize: `var(--typo-${value}-fz)`,
              },
            ];
          },
        },
        {
          values: {
            display: "display",
            headline: "headline",
            body: "body",
          },
        },
      );
```

**Comment by wongjn** [comment]
I would presume that calling `addBase()` inside `matchComponents()` is too "late" in the lifecycle of Tailwind's processes. I couldn't find any examples in the documentation of this either and since it doesn't work, I would assume it's not really *supposed* to work.

**Comment by adamwathan** [comment]
Hey! Converting this to a discussion because not a bug per se but still happy to chat it through and find a solution.

What you've got here isn't really the intended usage of these APIs, all of the functions like `addBase`, `addUtilities`, `matchUtilities`, etc. are all meant to be sort of "top-level" APIs, and invoking them in that `matchComponents` callback will lead that code to actually run too late so the styles are never properly registered.

Right now there's not really any way to sort of conditionally add styles to another layer when a plugin meant for a different layer runs, which I think is what you're trying to do if I'm understanding right.

This is probably what I would do if I'm understanding your code right:

https://play.tailwindcss.com/hOAu2zam47?file=config

At the same time though the whole thing feels very clever to me in a way I'd try and avoid. If you can explain the real UI/developer experience problem you're trying to solve maybe I can suggest a different approach.

**Comment by MatthiasDunker** [comment]
Thanks @wongjn 

Hey @adamwathan ,
thanks for your answer. You got me right :)
I am currently writing a CSS variable framework. The fun part is to have semantic composite classes (`.typo-display` for example) that can be modified by modifier classes (`.size-small` for example). I think of it as kind of layer styles (can be used for typography, colors, shapes...). Sticking to the `.typo-<x>` example, instead of adding specific utility classes and having to offer a lot of variants for different sizes, my idea is based on having a semantic schema expressed by one composite class like

```CSS
.typo-display {
  font-family: var(--typo-display-ff);
  font-size: var(--typo-display-fz);
  line-height: var(--typo-display-lh);
  font-weight: var(--typo-display-fw);
}
```
The CSS variables will have default styles on the root level.
But modifier classes are allowed to change them.

```CSS
.size-small {
  --typo-display-ff: "Calendas Plus";
  --typo-display-fz: 0.9rem;
  --typo-display-lh: 1.11111;
  --typo-display-fw: 400;
}
```
To keep it simple and enjoyable, I wanted to create integrations for Tailwindcss, Unocss etc. ... but is not that straight forward as I though.
And in the case of Tailwind, I thought the ideal way to go is to render a class only if it has been detected in the source files.
Therefore, the modifier class does not have to render a `.size-small {--typo-display-fz: 0.9rem; }` if a `.typo-display` has not been detected.

Btw. the framwork is called [Matricss](https://www.npmjs.com/package/@matricss/core) (very lean, no docs, and has to be tested a lot). Happy to tell more about it, if it has matured or interest is big enough :)
